### PR TITLE
Convert boxed primtive types to kotlin types

### LINKED ISSUE #139: Incorrect generation for boxed primitives when used as generics
If you read the type `SomeClass<Integer>` and write it you get `SomeClass<Integer>` rather than `SomeClass<Int>`. The generated code compiles but complains that `Integer` should not be used.

### PR DESCRIPTION
**Pull request by ansman**
This fixes #139

### CODE REVIEW

**Comment by ansman** [comment]
One outstanding question, should `java.lang.String` be converted to `kotlin.String`?

**Comment by ansman** [comment]
Also, I'm not sure about the location of these tests.

File: src/main/java/com/squareup/kotlinpoet/ClassName.kt [hunk 1]
```diff
@@ -203,6 +213,18 @@ class ClassName private constructor(
       }
       names += getPackage(this).qualifiedName.toString()
       names.reverse()
+      if (names.size == 2 && names[0] == "java.lang") {
+        when (names[1]) {
```
Comment (swankjesse) [inline src/main/java/com/squareup/kotlinpoet/ClassName.kt]: This seems like the wrong place to do this work. I think a more conventional mechanism would be for ClassName to have a method like `toKotlinClassName()` ?

Comment (swankjesse) [inline src/main/java/com/squareup/kotlinpoet/ClassName.kt]: ... also that’d make it possible to support other types, like MutableList?

Comment (ansman) [inline src/main/java/com/squareup/kotlinpoet/ClassName.kt]: Yeah, it felt a bit misplaced and I like an instance method.
That would make it less automatic which is fine as long as it's clearly documented I guess.

I wondered about things like MutableList but it's ambiguous if List should be MutableList or List, perhaps there is a good way to solve that though.

Comment (ansman) [inline src/main/java/com/squareup/kotlinpoet/ClassName.kt]: Oh, I see what you mean. Like you could have `toMutable` or something like that perhaps.

**Comment by ansman** [comment]
@swankjesse I added it as a function on `TypeName`. There are a couple of outstanding questions, I put some in the code but the other one is regarding the collection types.
Perhaps a `asMutable`/`asImmutable` function could work? It would just be a bit confusing to call that on `Int` for example.

File: src/main/java/com/squareup/kotlinpoet/ClassName.kt [hunk 2]
```diff
@@ -48,6 +48,23 @@ class ClassName private constructor(
 
   override fun asNonNullable() = ClassName(names, false, annotations)
 
+  override fun toKotlinType(): ClassName = when (packageName()) {
+    "java.lang" -> when (simpleName()) {
```
Comment (ansman) [inline src/main/java/com/squareup/kotlinpoet/ClassName.kt]: Are there more types that should be handled here?

Comment (JakeWharton) [inline src/main/java/com/squareup/kotlinpoet/ClassName.kt]: There's probably hundreds to the point where we're not going to be able to maintain a list.

**Comment by swankjesse** [comment]
Unclear what action to take on this. Closing for now.

**Comment by ansman** [comment]
@swankjesse Any reason for not merging it?

File: src/test/java/com/squareup/kotlinpoet/ClassNameTest.kt [hunk 3]
```diff
@@ -22,6 +22,7 @@ import org.junit.Assert.assertEquals
 import org.junit.Assert.fail
 import org.junit.Rule
 import org.junit.Test
+import javax.lang.model.type.TypeMirror
```
Comment (Egorand) [inline src/test/java/com/squareup/kotlinpoet/ClassNameTest.kt]: This import looks unused

Comment (ansman) [inline src/test/java/com/squareup/kotlinpoet/ClassNameTest.kt]: Fixed

**Comment by uluckas** [comment]
Is there any good reason not to fix the boxed types thing, while thinking about how to handle other kotlin types?

Here is one more usecase:
I try to generate an override for an existing method
```kotlin
    open fun getI(i: CompletableDeferred<Int>)
```

And due to this issue, my generated code will be:
```kotlin
    override fun getI(i: CompletableDeferred<Integer>)
```

which won't compile.

**Comment by JakeWharton** [comment]
That's a different issue. We should be reading the Kotlin type like the
compiler would.

**Pull request closed without merging**